# Capstone Project - InsaidTelecom

# Summarization of Learning

## 9 steps of EDA
1. Define Problem
2. Choose right tools
3. Collection of data
4. Pre-profile
5. Pre processing of data (Clean, remove unnecessary, add relevant data)
6. Post-profile
7. Ask right Questions
8. Conclusion or Summarization
9. Actionable Insights (low hanging fruits)


# Introduction

## Problem Statement 

**InsaidTelecom**, one of the leading telecom players, understands that customizing offering
is very important for its business to stay competitive.
Currently, InsaidTelecom is seeking to leverage behavioral data from more than 60% of the 50 million mobile devices active daily in India
to help its clients better understand and interact with their audiences.

In this **consulting assignment**, Insaidians are expected to build a dashboard to understand
user's demographic characteristics based on their mobile usage, geolocation, and mobile device properties.
Doing so will help millions of developers and brand advertisers around the world pursue
data-driven marketing efforts which are relevant to their users and catered to their preferences.

# Importing Packages

## Importing and Installing the packages

In [5]:
import numpy as np 

import pandas as pd
pd.set_option('mode.chained_assignment', None)                                    # To suppress pandas warnings.
pd.set_option('display.max_colwidth', None)                                       # To display all the data in each column
pd.set_option('display.max_columns', None)                                        # To display every column of the dataset in head()
pd.set_option('display.precision', 3)                                             # To display values only upto three decimal places.
pd.options.display.float_format = '{:.2f}'.format

import warnings
warnings.filterwarnings('ignore')                                                 # To suppress all the warnings in the notebook.

In [6]:
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns
sns.set(style='whitegrid', font_scale=1.3, color_codes=True)                      # To apply seaborn styles to the plots.

from datetime import datetime, timedelta

In [7]:
# To install pandas profiling please run this command.
#while re-running the notebook we dont want the pip command, so commenting it out
#!pip install pandas-profiling --upgrade
#!pip uninstall !pip uninstall pandas_profiling
#!pip install pandas_profiling[notebook,html]

In [8]:
from pandas_profiling import ProfileReport

## Handling warnings

In [9]:
warnings.filterwarnings('ignore') # ignore warnings.
%config IPCompleter.greedy = True # autocomplete feature.
pd.options.display.max_rows = None # set maximum rows that can be displayed in notebook.
pd.options.display.max_columns = None # set maximum columns that can be displayed in notebook.
pd.options.display.precision = 2 # set the precision of floating point numbers.

# Loading Data

## Description of Data

In this assignment, we are going to study the demographics of a user (gender and age) based on their app download and usage behaviors.

The Data is collected from mobile apps that use InsaidTelecom services. Full recognition and consent from individual user of those apps have been obtained, and appropriate anonymization have been performed to protect privacy. Due to confidentiality, we won't provide details on how the gender and age data was obtained.

The data schema can be represented in the following table:

**gender_age_train** - Devices and their respective user gender, age and age_group

**phone_brand_device_model** - device ids, brand, and models phone_brand: note that few brands are in Chinese

**events_data** - when a user uses mobile on INSAID Telecom network, the event gets logged in this data.

Each event has an event id, location (lat/long), and the event corresponds to frequency of mobile usage.
**timestamp**: when the user is using the mobile.

In [10]:
df = pd.read_csv("C://Users//jeete//OneDrive//Desktop//Kasturi//Capstone Project//events_data.csv")
df.head()

event_id               device_id            timestamp  longitude  latitude  \
0   2765368  2973347786994914304.00  2016-05-07 22:52:05      77.23     28.73   
1   2955066  4734221357723754496.00  2016-05-01 20:44:16      88.39     22.66   
2    605968 -3264499652692492800.00  2016-05-02 14:23:04      77.26     28.76   
3    448114  5731369272434022400.00  2016-05-03 13:21:16      80.34     13.15   
4    665740   338888002570799936.00  2016-05-06 03:51:05      86.00     23.84   

       city       state  
0     Delhi       Delhi  
1  Calcutta  WestBengal  
2     Delhi       Delhi  
3   Chennai   TamilNadu  
4    Bokaro   Jharkhand

In [12]:
#to identify null, missing data and the type of Data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3252950 entries, 0 to 3252949
Data columns (total 7 columns):
 #   Column     Dtype  
---  ------     -----  
 0   event_id   int64  
 1   device_id  float64
 2   timestamp  object 
 3   longitude  float64
 4   latitude   float64
 5   city       object 
 6   state      object 
dtypes: float64(3), int64(1), object(3)
memory usage: 173.7+ MB


<span style='background:orange'>Observation:There are 3 float type data, 1 integer type and 3 object type data

In [51]:
# to find the shape of the data, though seen in the .info() , double checking here
df.shape

(3252950, 7)

<span style='background:orange'>Observation:As seen above in .info() we have 3252950 rows and 7 columns</span>

In [13]:
# to check for the null values. Already checked in .info(), double checking here
df.isnull().sum()

event_id       0
device_id    453
timestamp      0
longitude    423
latitude     423
city           0
state        377
dtype: int64

<span style='background:orange'> Observation: There are null values in device_id,  longitude, latitude and state column </span>

In [14]:
#to check the statistical data like count, mean, std, min, max, IQR
df.describe()

event_id               device_id  longitude   latitude
count 3252950.00              3252497.00 3252527.00 3252527.00
mean  1626475.50   101220009585509024.00      78.16      21.69
std    939045.92  5316758188197051392.00       4.24       5.79
min         1.00 -9222956879900150784.00      12.57       8.19
25%    813238.25 -4540611333857475072.00      75.84      17.80
50%   1626475.50   172682011159278816.00      77.27      22.16
75%   2439712.75  4861813234983622656.00      80.32      28.68
max   3252950.00  9222849349208140800.00      95.46      41.87

# Loading Data Using SQL

In [36]:
# Importing sql connector
import mysql.connector

In [37]:
# Installing mysql-connector to connect to database

!pip install mysql-connector

In [38]:
# Connecting to mysql database

mydb = mysql.connector.connect(host='cpanel.insaid.co', user='student', passwd='student', 
                                                          database='Capstone1')

In [39]:
mydb

In [40]:
# Creating the cursor to point each instance of the database

my_cursor = mydb.cursor()

In [41]:
# Fetching all the data from gender_age_train table using cursor object

my_cursor.execute('select * from gender_age_train')

In [42]:
# Store the data available in cursor object in gender_age_train variable

gender_age_train = my_cursor.fetchall()

In [43]:

# Converting gender_age_train variable into a Dataframe

gender_age_train = pd.DataFrame(gender_age_train)

In [44]:
gender_age_train.head()

0  1   2       3
0 -8076087639492063270  M  35  M32-38
1 -2897161552818060146  M  35  M32-38
2 -8260683887967679142  M  35  M32-38
3 -4938849341048082022  M  30  M29-31
4   245133531816851882  M  30  M29-31

In [45]:
gender_age_train.info()  # There is no missing information in gender_age_train dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       74645 non-null  int64 
 1   1       74645 non-null  object
 2   2       74645 non-null  int64 
 3   3       74645 non-null  object
dtypes: int64(2), object(2)
memory usage: 2.3+ MB


In [46]:
# Fetching all the data from phone_brand_device_model table using cursor object

my_cursor.execute('select * from phone_brand_device_model')

In [47]:
# Store the data available in cursor object in phone_brand_device_model variable

phone_brand_device_model = my_cursor.fetchall()

In [48]:
# Converting phone_brand_device_model variable into a Dataframe

phone_brand_device_model = pd.DataFrame(phone_brand_device_model)

In [49]:

phone_brand_device_model.head()

0     1        2
0  1877775838486905855  vivo      Y13
1 -3766087376657242966    小米     V183
2 -6238937574958215831  OPPO      R7s
3  8973197758510677470    三星    A368t
4 -2015528097870762664    小米  红米Note2

In [50]:
phone_brand_device_model.info()  # There is no missing information in phone_brand_device_model

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87726 entries, 0 to 87725
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       87726 non-null  int64 
 1   1       87726 non-null  object
 2   2       87726 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.0+ MB
